revokefs/writer.c

```diff
@@ -0,0 +1,876 @@
+
+  /* Note we can't use AT_SYMLINK_NOFOLLOW yet;
+   * https://marc.info/?l=linux-kernel&m=148830147803162&w=2
+   * https://marc.info/?l=linux-fsdevel&m=149193779929561&w=2
+   */
+  if (fchmodat (basefd, path, mask_mode (mode), 0) != 0)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_chmod(int writer_socket, const char *path, mode_t mode)
+{
+  return request_path_int (writer_socket, REVOKE_FS_CHMOD, path, mode);
+}
+
+static ssize_t
+handle_chown (RevokefsRequest *request,
+              gsize data_size,
+              RevokefsResponse *response)
+{
+  g_autofree char *path = get_valid_path (request->data, data_size);
+  uid_t uid = request->arg1;
+  gid_t gid = request->arg2;
+
+  if (fchownat (basefd, path, uid, gid, AT_SYMLINK_NOFOLLOW) != 0)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_chown(int writer_socket, const char *path, uid_t uid, gid_t gid)
+{
+  return request_path_int_int (writer_socket, REVOKE_FS_CHOWN, path, uid, gid);
+}
+
+static ssize_t
+handle_truncate (RevokefsRequest *request,
+                 gsize data_size,
+                 RevokefsResponse *response)
+{
+  g_autofree char *path = get_valid_path (request->data, data_size);
+  off_t size = request->arg1;
+
+  glnx_autofd int fd = openat (basefd, path, O_NOFOLLOW|O_WRONLY);
+  if (fd == -1)
+    response->result = -errno;
+  else
+    {
+      if (ftruncate (fd, size) == -1)
+        response->result = -errno;
+      else
+        response->result = 0;
+    }
+
+  return 0;
+}
+
+int
+request_truncate (int writer_socket, const char *path, off_t size)
+{
+  return request_path_i64_i64 (writer_socket, REVOKE_FS_TRUNCATE, path, size, 0);
+}
+
+static ssize_t
+handle_utimens (RevokefsRequest *request,
+                gsize data_size,
+                RevokefsResponse *response)
+{
+  g_autofree char *path = NULL;
+  struct timespec *tv;
+
+  if (request->arg1 + sizeof (struct timespec) * 2 != data_size)
+    {
+      g_printerr ("Invalid data size\n");
+      exit (1);
+    }
+
+  path = get_valid_path (request->data, request->arg1);
+  tv = (struct timespec *)(request->data + request->arg1);
+
+  if (utimensat (basefd, path, tv, AT_SYMLINK_NOFOLLOW) == -1)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_utimens (int writer_socket, const char *path, const struct timespec tv[2])
+{
+  return request_path_data (writer_socket, REVOKE_FS_UTIMENS, path,
+                            (guchar *)tv, sizeof (struct timespec) * 2);
+}
+
+static ssize_t
+handle_open (RevokefsRequest *request,
+             gsize data_size,
+             RevokefsResponse *response)
+{
+  g_autofree char *path = get_valid_path (request->data, data_size);
+  int mode = request->arg1;
+  int flags = request->arg2;
+  int fd;
+
+
+  /* We need to specially handle O_TRUNC */
+  fd = openat (basefd, path, flags & ~O_TRUNC, mask_mode (mode));
```

Comment:
Similar issues here with symlinks. We can't allow you to create a symlink to anywhere and then open it writable. I *believe* fuse itself will resolve a symlink if it is opened, so unless the app explicitly tries to trick things by racing with the backend we should never actually get a path to a symlink here. This means we should just always add O_NOFOLLOW to flags here. However, that theory does need verification.